### ISSUE TITLE
CUDA multinomial with replacement can select zero-probability events

### ISSUE THREAD

**Issue by coventry** [pytorch/pytorch#4858]
I'm running Ubuntu 16.04.3 on an AWS P3.2xlarge, with the NVIDIA libcuda1-384 package, version 384.111-0ubuntu0.16.04.1, and pytorch version 0.3.0.post4 and cuda90, installed via conda.

In some circumstances, the CUDA multinomial sampler with `replacement=True` can sample an event with zero probability.

The script below reproduces this. I haven't tried to reduce the distribution to a minimal case.  The output is below the script.

I've attached [the RNG state](https://github.com/pytorch/pytorch/files/1665704/failed_state.pt.gz) just prior to the failure. Reproduction from this state is demonstrated in the script.

Bests regards,
Alex

```python
import torch as T
import os

freqs = T.cuda.FloatTensor([
    0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
    0.03178183361887932, 0.027680952101945877, 0.033176131546497345,
    0.046052902936935425, 0.07742464542388916, 0.11543981730937958,
    0.14148041605949402, 0.15784293413162231, 0.13180233538150787,
    0.08271478116512299, 0.049702685326337814, 0.027557924389839172,
    0.018125897273421288, 0.011851548217236996, 0.010252203792333603,
    0.007422595750540495, 0.005372154992073774, 0.0045109698548913,
    0.0036087757907807827, 0.0035267581697553396, 0.0018864056328311563,
    0.0024605290964245796, 0.0022964938543736935, 0.0018453967059031129,
    0.0010662291897460818, 0.0009842115687206388, 0.00045109697384759784,
    0.0007791675161570311, 0.00020504408166743815, 0.00020504408166743815,
    0.00020504408166743815, 0.00012302644609007984, 0.0,
    0.00012302644609007984, 4.100881778867915e-05, 0.0, 0.0, 0.0, 0.0,
    0.0, 0.0])

for i in range(1_000_000_000):
    state = T.cuda.get_rng_state()
    sample = T.multinomial(freqs, 1000, True)
    if freqs[sample].min() == 0:
        break

print(72*'-')
print(f'failure after {i} iterations')
sample_idx = (freqs[sample] == 0).nonzero()[0][0]
sampled = sample[sample_idx]
print(f'{sample_idx}th element of last sample was {sampled}, '
      f'which has probability {freqs[sampled]}')

print(72*'-')
SAVE_PATH = 'failed_state.pt'
T.save(state, open(SAVE_PATH, 'wb'))
print(f'\nRNG state saved to `{SAVE_PATH}`')

T.cuda.set_rng_state(T.load(open(SAVE_PATH, 'rb')))
assert freqs[T.multinomial(freqs, 1000, True)].min() == 0

print(72*'-')
print(f'\npytorch version: {T.__version__}\n')
print(72*'-')
print('Operating system:')
os.system('lsb_release -a')
print(72*'-')
print('\nGPU information:')
os.system('nvidia-smi')
print(72*'-')
print('\nCUDA library information:')
os.system('dpkg -l | grep -i cuda')
```

```
ubuntu@ip-172-30-1-60:~/workdir$ python -i fail.py 
------------------------------------------------------------------------
failure after 4309 iterations
631th element of last sample was 0, which has probability 0.0
------------------------------------------------------------------------

RNG state saved to `failed_state.pt`
------------------------------------------------------------------------

pytorch version: 0.3.0.post4

------------------------------------------------------------------------
Operating system:
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.3 LTS
Release:        16.04
Codename:       xenial
------------------------------------------------------------------------

GPU information:
Thu Jan 25 21:17:03 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.111                Driver Version: 384.111                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   38C    P0    46W / 300W |   4558MiB / 16152MiB |     31%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0       988      C   python                                       738MiB |
|    0     15193      C   python                                      3810MiB |
+-----------------------------------------------------------------------------+
------------------------------------------------------------------------

CUDA library information:
ii  libcuda1-384                                    384.111-0ubuntu0.16.04.1     amd64                        NVIDIA CUDA runtime library
```

**Comment by SsnL** [comment]
Thanks for the detailed report. The bug is not reproducible on master with the provided check point. It's likely fixed with the many changes done to multinomial.

**Comment by SsnL** [comment]
Hold on. I reproduced it.

**Comment by wouterkool** [comment]
I did not know this issue was open and also created a case [here](https://discuss.pytorch.org/t/bad-behavior-of-multinomial-function/10232/3). If it is any help, I found that it depends on the range of the values as well, if I change logits_range in the script below to 1 or 100 it does not happen.

```
import torch
from torch.autograd import Variable

import torch.nn.functional as F
import numpy as np
from tqdm import tqdm


def test(n, hot=False, logits_range=10):
    torch.manual_seed(1234)

    logits = Variable(torch.Tensor(128, 50).uniform_(-logits_range, logits_range).cuda())

    # Set randomly 40 elements per row to 0
    mask = torch.zeros_like(logits).byte()
    _, idx_mask = Variable(torch.Tensor(128, 50).uniform_(0, 1).cuda()).topk(40, 1)
    mask.scatter_(1, idx_mask, True)

    logits[mask] = -np.inf

    probs = F.softmax(logits, dim=1)

    assert (probs[mask] == 0).all()
    assert (torch.abs(probs.sum(1) - 1) < 1e-6).all()

    if hot:
        with open('rng_state.pt', 'rb') as f:
            rng_state = torch.load(f)
        torch.cuda.set_rng_state(rng_state)

    for j in tqdm(range(n)):

        rng_state = torch.cuda.get_rng_state()

        sample = probs.multinomial(1).squeeze(-1)
        mask_sample = mask.gather(1, sample.unsqueeze(-1)).squeeze(-1)

        if mask_sample.any():
            print("Sampled value that was masked and had probability 0 in iteration {}".format(j))
            wrong = torch.nonzero(mask_sample).squeeze(-1)
            print("Wrong samples: indices {}, sampled {}, probs {}".format(
                wrong.data.cpu().numpy().tolist(),
                sample[wrong].data.cpu().numpy().tolist(),
                probs[wrong, sample[wrong]].data.cpu().numpy().tolist()
            ))

            if hot:
                break

            with open('rng_state.pt', 'wb') as f:
                torch.save(rng_state, f)


if __name__ == "__main__":
    with torch.cuda.device(0):
        test(100000, hot=False)
```

**Comment by t-vi** [comment]
Hi,

thank you @coventry for the test case. I think I found the bug, your code was invaluable in finding and testing a fix.

Best regards

Thomas

P.S.: Please ignore the suboptimal workarounds before the pull-request.

**Comment by SsnL** [comment]
@soumith this can be closed

**Comment by rfejgin** [comment]
I have seen this happen with PyTorch 0.4.0.
Interestingly, the zero-probability event that was select happened to be index 0 out of 255, even though there were many other zero-probability indices - so index zero was chosen may be a clue?
In my case the number of samples (n_samples) was 1, which seems to map to this optimized flavor  'sampleMultinomialOnce()'.
I have not been able to consistently reproduce this so far.

**Comment by t-vi** [comment]
I'm still kind of proud for debugging and fixing this (one of my early deep into CUDA patches and I did most of the searching while on the train with a sketchy connection to the GPU computer), but I only improved the binary search based multi-item case, not the sampleMultinomialOnce.

This reproduces the problem for me:
```
import torch
a = torch.zeros(3421,2, device="cuda")
a[:,1] = 1
torch.cuda.manual_seed(5214)
b = torch.multinomial(a, 1)
assert b.min().item()>0
```
I can take a look, but I herd whispers that the random functions might be in for a rework anyway.

**Comment by rfejgin** [comment]
Yes, I can reproduce the issue with your code. I was also able to create my own reproducible test case, but yours is simpler so let's stick with that.

@soumith @apaszke can we reopen this bug?

**Comment by rfejgin** [comment]
It looks like the problem is broader than we initially thought. Zero probability events can get selected anywhere in the distribution, not only just if they are the first or last bin. The script below reproduces the problem in bins in the middle of the distribution the in just a few seconds. 

... 

```
import sys
import torch

batch_size = 1024
dist_size = 2048
torch.manual_seed(1)
torch.cuda.manual_seed(1)
weights = torch.zeros(batch_size, dist_size).cuda()
with torch.no_grad():
    k = 0
    while True:
        if k % 10 == 0:
            print("iteration %d" % k)
        # create a different unnormalized PDF each time.
        weights.uniform_(0.4, 0.6) # use rather large weights

        # zero out about half of the probabilities
        cond = torch.rand(batch_size, dist_size).ge(0.5)
        weights[cond] = 0
        assert(weights.sum(dim=1).gt(0).all().item()) # make sure we didn't accidentally zero out all the weights in a distribution

        # Sample
        s = weights.multinomial(1).squeeze(1)

        # Check if any selected samples had zero probability
        selected_probs = weights[torch.arange(batch_size, dtype=torch.long), s]
        if not selected_probs.eq(0).sum() == 0:
            # Bug detected. Print out some debug info.
            print("\nError: Zero-probability event sampled!")
            for i in range(batch_size):
                if selected_probs[i] == 0:
                        print("bin %d out of %d in distribution %d was chosen even though its probability is %.9f" %(s[i], dist_size, i, selected_probs[i]))
            sys.exit(1)
        k = k + 1
```
**Output:**
```
iteration 0
iteration 10
iteration 20
iteration 30
iteration 40
iteration 50
iteration 60
iteration 70
iteration 80

Error: Zero-probability event sampled!
bin 1679 out of 2048 in distribution 461 was chosen even though its probability is 0.000000000
```

**Comment by rfejgin** [comment]
Another finding from the thread for @t-vi's pull request  (https://github.com/pytorch/pytorch/pull/9960) is that there appears to be a problem with some of the floating point calculations in the code.

For example, I added a this printout:
```printf("Zero probability selected! cat=%d, probability=%.9g, prevBucket=%.9g, curBucket=%.9g\n", cat, prob, prevBucket, curBucket);```
at this line:
https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCTensorRandom.cuh#L263

and, when running the above script, got the output:
```Zero probability selected! cat=1679, probability=0, prevBucket=0.811896205, curBucket=0.811896265```

Note how ```prevBucket``` and ```curBucket```, which should be equal since this is a zero-probability bin, are in fact slightly different. This leads to the bin getting selected even though it has zero probability because this condition evaluates as true:
https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCTensorRandom.cuh#L257

Assuming that such small drifts in floating point calculations are hard to avoid, maybe the condition should be made to work with a tolerance rather than a hard ```>``` check? For example, we could select the bin only if curBucket > prevBucket + eps.

@t-vi @apaszke @soumith - Thoughts?

**Comment by rfejgin** [comment]
oh, and just wanted to thank @t-vi for the fix candidates and great discussion!

**Comment by rfejgin** [comment]
Specifically, we could do something like this:
Change https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCTensorRandom.cuh#L257
to 
```      bool inBucket =
        (cat < categories) &&
        (!THCNumerics<T>::gt(sample, curBucket)) &&
        (THCNumerics<T>::gt(sample, prevBucket)) &&
        (THCNumerics<T>::gt(THCNumerics<T>::sub(curBucket, prevBucket), ScalarConvert<float, T>::to(1e-07)));```
except that the threshold (1e-7 in this example) would need to depend on the data type T.

**Comment by t-vi** [comment]
I'd still think we should be able to take it apart to know why it fails here. I'll give it another shot with debugging - hopefully also finding out why the same kernel seems to be behaving differently in the extension and in THC.

**Comment by ailzhang** [comment]
@t-vi  Any progress on this?

**Comment by t-vi** [comment]
The PR fixes it.

**Comment by HarveyYan** [comment]
the problem is back on torch 2.0.0. Categorical samples positions whose logits have been masked to -inf. Only happens on GPUs.

**Comment by josephdviviano** [comment]
I can confirm that this bug is back. 

It is reported early this year in https://github.com/pytorch/pytorch/issues/91863
and https://github.com/pytorch/pytorch/issues/100884, and is causing an issue with our gflownet library https://github.com/saleml/torchgfn/issues/51 which depends on actions sampled always being "valid" (probability > 0).

@saleml @vict0rsch
